base/show.jl

```diff
@@ -487,37 +484,256 @@ function io_has_tvar_name(io::IOContext, name::Symbol, @nospecialize(x))
+        end
+    end
+end
+
+function show_typealias(io::IO, x::Type)
+    properx = makeproper(io, x)
+    alias = make_typealias(properx)
+    alias === nothing && return false
+    show_typealias(io, alias[1], x, alias[2])
+    return true
+end
+
+function make_typealiases(x::Type)
+    Any <: x && return Core.svec(), Union{}
+    x <: Tuple && return Core.svec(), Union{}
+    mods = modulesof!(Set(), x)
+    Core in mods && push!(mods, Base)
+    aliases = SimpleVector[]
+    vars = Dict{Symbol,TypeVar}()
+    xenv = UnionAll[]
+    for p in uniontypes(unwrap_unionall(x))
+        p isa UnionAll && push!(xenv, p)
+    end
+    x isa UnionAll && push!(xenv, x)
+    for mod in mods
+        for name in names(mod)
+            if isdefined(mod, name) && !isdeprecated(mod, name) && isconst(mod, name)
+                alias = getfield(mod, name)
+                if alias isa Type && !has_free_typevars(alias) && !isvarargtype(alias) && !print_without_params(alias) && !(alias <: Tuple)
+                    (ti, env) = ccall(:jl_type_intersection_with_env, Any, (Any, Any), x, alias)::SimpleVector
+                    ti === Union{} && continue
+                    mod in modulesof!(Set(), alias) || continue # make sure this alias wasn't from an unrelated part of the Union
+                    env = env::SimpleVector
+                    applied = isempty(env) ? alias : alias{env...}
+                    ul = unionlen(applied)
+                    for p in xenv
+                        applied = rewrap_unionall(applied, p)
+                    end
+                    has_free_typevars(applied) && continue
+                    applied <: x || continue # parameter matching didn't make a subtype
+                    print_without_params(x) && (env = Core.svec())
+                    push!(aliases, Core.svec(GlobalRef(mod, name), env, applied, (ul, -length(env))))
+                end
+            end
+        end
+    end
+    if isempty(aliases)
+        return Core.svec(), Union{}
+    end
+    sort!(aliases, by = x -> x[4], rev = true) # heuristic sort by "best" environment
+    let applied = Union{}
+        applied1 = Union{}
+        keep = SimpleVector[]
+        prev = (0, 0)
+        for alias in aliases
+            if alias[4][1] < 2
+                if !(alias[3] <: applied)
+                    applied1 = Union{applied1, alias[3]}
+                    push!(keep, alias)
+                end
+            elseif alias[4] == prev || !(alias[3] <: applied)
+                applied = applied1 = Union{applied1, alias[3]}
+                push!(keep, alias)
+                prev = alias[4]
+            end
+        end
+        return keep, applied1
+    end
+end
+
+function show_unionaliases(io::IO, x::Union)
+    properx = makeproper(io, x)
+    aliases, applied = make_typealiases(properx)
+    first = true
+    for typ in uniontypes(x)
+        if !isa(typ, TypeVar) && rewrap_unionall(typ, properx) <: applied
+            continue
```

Comment:
We should avoid changing the print order of the union elements.

Reply:
Hm, that'd require some additional complexity, since in full generality it's not possible. Is it that bad to split it into two independently partially-sorted sets in this way?